### PR TITLE
feature: Cache Adapters and REST Manager

### PR DESCRIPTION
**Pull request by DjDeveloperr**
In this PR, I have:
- Implemented CacheAdapters. They're flexible, customizable, classes which implement ICacheAdapter interface, allowing user to implement caching, anywhere. There are two adapters shipped with this commit, one is DefaultCacheAdapter which caches data in Collections.
Note: Now API-Payloads are cached, not the dynamic objects for Redis compatibility. 
- RESTManager. Requests Managed in one place, including Rate Limit handling. All requests are done from here (unless I forgot some class)
- And I've made few improvements in other files.
However, I had to remove some things to make new things work, like roles and members of Guild aren't cached at the moment, which I'll be adding with future PRs.

### CODE REVIEW

File: src/gateway/handlers/messageCreate.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import { MessagePayload } from "../../types/channelTypes.ts"
+import { Gateway, GatewayEventHandler } from '../index.ts'
+
+export const messageCreate: GatewayEventHandler = async(
+  gateway: Gateway,
+  d: MessagePayload
+) => {
+  let channel = await gateway.client.channels.get(d.channel_id)
+  // Fetch the channel if not cached
+  if(!channel) channel = (await gateway.client.channels.fetch(d.channel_id) as any) as Channel
```
Comment (Helloyunho) [inline src/gateway/handlers/messageCreate.ts:14]: channels from event `messageCreate` will always be `TextChannel`

File: src/structures/guild.ts [hunk 2]
```diff
@@ -28,7 +29,7 @@ export class Guild extends Base {
   verificationLevel?: string
   defaultMessageNotifications?: string
   explicitContentFilter?: string
-  roles?: Role[]
+  roles: RolesManager = new RolesManager(this.client, this)
   emojis?: Emoji[]
```
Comment (Helloyunho) [inline src/structures/guild.ts:33]: This should be an `EmojiManager`

Comment (Helloyunho) [reply src/structures/guild.ts:33]: members and channels properties too

**Review by Helloyunho — CHANGES_REQUESTED** [review]
Please fix these

File: src/gateway/handlers/messageCreate.ts (hunk 1 above)
Comment (DjDeveloperr) [reply src/gateway/handlers/messageCreate.ts]: Fixed.

File: src/structures/guild.ts (hunk 2 above)
Comment (DjDeveloperr) [reply src/structures/guild.ts:33]: These properties are yet to be implemented using "Child Manager" classes which I'll be making using future PRs.

**Review by Helloyunho — APPROVED** [review]
NoiCe

File: src/structures/guild.ts (hunk 2 above)
Comment (Helloyunho) [reply src/structures/guild.ts:33]: okay

**Pull request merged**